atc/worker/placement.go

```diff
@@ -27,269 +33,400 @@ func (err NoWorkerFitContainerPlacementStrategyError) Error() string {
-			logger.Info("worker-busy")
-			continue
-		}
+		candidates = append(candidates, worker)
+		taskCounts[worker] = activeTasks
+	}
+
+	sort.SliceStable(candidates, func(i, j int) bool {
+		return taskCounts[candidates[i]] < taskCounts[candidates[j]]
+	})
+
+	return candidates, nil
+}
+
+func (strategy *LimitActiveTasksStrategy) Pick(logger lager.Logger, worker Worker, spec ContainerSpec) error {
+	if spec.Type != db.ContainerTypeTask {
+		return nil
+	}
+
+	activeTasks, err := worker.IncreaseActiveTasks()
+
+	if err != nil {
+		return err
+	}
 
-		workersByWork[activeTasks] = append(workersByWork[activeTasks], w)
-		if minActiveTasks == -1 || activeTasks < minActiveTasks {
-			minActiveTasks = activeTasks
+	if strategy.maxTasks > 0 && activeTasks > strategy.maxTasks {
+		_, err := worker.DecreaseActiveTasks()
+		if err != nil {
+			logger.Error("failed-to-decrease-active-tasks", err)
 		}
+
+		return ErrTooManyActiveTasks
 	}
 
-	return workersByWork[minActiveTasks], nil
+	return nil
 }
 
-func (strategy *LimitActiveTasksPlacementStrategyNode) ModifiesActiveTasks() bool {
-	return true
+func (strategy *LimitActiveTasksStrategy) Release(logger lager.Logger, worker Worker, spec ContainerSpec) {
+	if spec.Type != db.ContainerTypeTask {
+		return
+	}
+
+	_, err := worker.DecreaseActiveTasks()
+	if err != nil {
+		logger.Error("failed-to-decrease-active-tasks", err)
+	}
 }
 
-func (strategy *LimitActiveTasksPlacementStrategyNode) StrategyName() string {
-	return strategy.GivenName
+func (strategy *LimitActiveTasksStrategy) Name() string {
+	return strategy.name
 }
 
-type LimitActiveContainersPlacementStrategyNode struct {
-	GivenName     string
+type LimitActiveContainersStrategy struct {
+	name          string
 	maxContainers int
 }
 
-func newLimitActiveContainersPlacementStrategy(name string, maxContainers int) ContainerPlacementStrategyChainNode {
-	return &LimitActiveContainersPlacementStrategyNode{
-		GivenName:     name,
+func newLimitActiveContainersStrategy(name string, maxContainers int) ContainerPlacementStrategy {
+	return &LimitActiveContainersStrategy{
+		name:          name,
 		maxContainers: maxContainers,
 	}
 }
 
-func (strategy *LimitActiveContainersPlacementStrategyNode) Choose(logger lager.Logger, workers []Worker, spec ContainerSpec) ([]Worker, error) {
-	candidates := []Worker{}
+func (strategy *LimitActiveContainersStrategy) Order(logger lager.Logger, workers []Worker, spec ContainerSpec) ([]Worker, error) {
+	candidates := append([]Worker(nil), workers...)
+	counts := make(map[Worker]int, len(candidates))
 
-	for _, w := range workers {
-		if strategy.maxContainers == 0 || w.ActiveContainers() <= strategy.maxContainers {
-			candidates = append(candidates, w)
-		}
+	for _, worker := range candidates {
+		counts[worker] = worker.ActiveContainers()
 	}
 
+	sort.SliceStable(candidates, func(i, j int) bool {
+		return counts[candidates[i]] < counts[candidates[j]]
```

Comment:
Consider the following placement strategy:
```
limit-active-containers,volume-locality
```

with max active containers is set to 100.

`worker1` has 50 containers, and `worker2` has 60 containers.
`worker1` has 1 volume locally, and `worker2` has 2 volumes locally.

Prior to this PR, Concourse would schedule on `worker2` - since both workers are both under the container limit, we resort to effectively only considering `volume-locality`. Since the `limit-active-containers` sort now takes precedence (which I think is what we want in general - the further up the chain, the more precedence), we'll pick `worker1` since it has fewer containers locally.

I think this can be fixed by not sorting on the number of containers, but sorting on whether the worker exceeds the number of containers or not (since `limit-active-containers` doesn't care about giving you the worker with the fewest containers active - that's what `fewest-build-containers` is for). e.g. something like:

```go
	exceedsThreshold := func(count int) int {
		if strategy.maxContainers > 0 && count > strategy.maxContainers {
			return 1
		} else {
			return 0
		}
	}
	sort.SliceStable(candidates, func(i, j int) bool {
		return exceedsThreshold(counts[candidates[i]]) < exceedsThreshold(counts[candidates[j]])
	}
```

So, we consider all counts less than the threshold to be 0, and all counts greater to be 1 - this way, we partition the slice while maintaining the original ordering.

**Note:** The same applies for `limit-active-volumes`, but **not** for `limit-active-tasks` (it's a weird one that both filters based on exceeding the threshold AND tries to give you the worker with the fewest tasks)